From cb960e29a694642eeceb7daca82a80ec021366d2 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Robert-Andr=C3=A9=20Mauchin?= <zebob.m@gmail.com>
Date: Sun, 14 Apr 2024 11:09:20 +0200
Subject: [PATCH] Compatibility with FFMPEG 7.0

channel_layout has been replace with ch_layout

avcodec_close has been deprecated in favor of avcodec_free_context

Fix #57
---
 a52/pcm_a52.c           | 68 ++++++++++++++++++++++++++++++-----------
 rate-lav/rate_lavrate.c | 30 +++++++++++++++---
 2 files changed, 77 insertions(+), 21 deletions(-)

diff --git a/a52/pcm_a52.c b/a52/pcm_a52.c
index 9da8897..66f3bb2 100644
--- a/a52/pcm_a52.c
+++ b/a52/pcm_a52.c
@@ -78,6 +78,11 @@
 #define av_frame_free avcodec_free_frame
 #endif
 
+#if LIBAVCODEC_VERSION_INT >= AV_VERSION_INT(61, 3, 100)
+#define USE_AVCODEC_FREE_CONTEXT
+#define USE_CH_LAYOUT
+#endif
+
 struct a52_ctx {
 	snd_pcm_ioplug_t io;
 	snd_pcm_t *slave;
@@ -628,7 +633,11 @@ static int a52_stop(snd_pcm_ioplug_t *io)
 static void a52_free(struct a52_ctx *rec)
 {
 	if (rec->avctx) {
-		avcodec_close(rec->avctx);
+		#ifdef USE_AVCODEC_FREE_CONTEXT
+			avcodec_free_context(&rec->avctx);
+		#else
+			avcodec_close(rec->avctx);
+		#endif
 		av_free(rec->avctx);
 		rec->avctx = NULL;
 	}
@@ -667,19 +676,35 @@ static void a52_free(struct a52_ctx *rec)
 static void set_channel_layout(snd_pcm_ioplug_t *io)
 {
 	struct a52_ctx *rec = io->private_data;
-	switch (io->channels) {
-	case 2:
-		rec->avctx->channel_layout = AV_CH_LAYOUT_STEREO;
-		break;
-	case 4:
-		rec->avctx->channel_layout = AV_CH_LAYOUT_QUAD;
-		break;
-	case 6:
-		rec->avctx->channel_layout = AV_CH_LAYOUT_5POINT1;
-		break;
-	default:
-		break;
-	}
+	#ifdef USE_CH_LAYOUT
+		switch (io->channels) {
+			case 2:
+				rec->avctx->ch_layout.u.mask = AV_CH_LAYOUT_STEREO;
+				break;
+			case 4:
+				rec->avctx->ch_layout.u.mask = AV_CH_LAYOUT_QUAD;
+				break;
+			case 6:
+				rec->avctx->ch_layout.u.mask = AV_CH_LAYOUT_5POINT1;;
+				break;
+			default:
+				break;
+		}
+	#else
+		switch (io->channels) {
+			case 2:
+				rec->avctx->channel_layout = AV_CH_LAYOUT_STEREO;
+				break;
+			case 4:
+				rec->avctx->channel_layout = AV_CH_LAYOUT_QUAD;
+				break;
+			case 6:
+				rec->avctx->channel_layout = AV_CH_LAYOUT_5POINT1;
+				break;
+			default:
+				break;
+		}
+	#endif
 }
 #else
 #define set_channel_layout(io) /* NOP */
@@ -695,8 +720,13 @@ static int alloc_input_buffer(snd_pcm_ioplug_t *io)
 	rec->frame->nb_samples = rec->avctx->frame_size;
 #if LIBAVCODEC_VERSION_INT >= AV_VERSION_INT(58, 91, 0)
 	rec->frame->format = rec->avctx->sample_fmt;
-	rec->frame->channels = rec->avctx->channels;
-	rec->frame->channel_layout = rec->avctx->channel_layout;
+	#ifdef USE_CH_LAYOUT
+		rec->frame->ch_layout.nb_channels = rec->avctx->ch_layout.nb_channels;
+		rec->frame->ch_layout.u.mask = rec->avctx->ch_layout.u.mask;
+	#else
+		rec->frame->channels = rec->avctx->channels;
+		rec->frame->channel_layout = rec->avctx->channel_layout;
+	#endif
 	if (av_frame_get_buffer(rec->frame, 0))
 		return -ENOMEM;
 #else
@@ -731,7 +761,11 @@ static int a52_prepare(snd_pcm_ioplug_t *io)
 
 	rec->avctx->bit_rate = rec->bitrate * 1000;
 	rec->avctx->sample_rate = io->rate;
-	rec->avctx->channels = io->channels;
+	#ifdef USE_CH_LAYOUT
+		rec->avctx->ch_layout.nb_channels = io->channels;
+	#else
+		rec->avctx->channels = io->channels;
+	#endif
 	rec->avctx->sample_fmt = rec->av_format;
 
 	set_channel_layout(io);
diff --git a/rate-lav/rate_lavrate.c b/rate-lav/rate_lavrate.c
index f78eea5..23a5294 100644
--- a/rate-lav/rate_lavrate.c
+++ b/rate-lav/rate_lavrate.c
@@ -18,11 +18,26 @@
 #include <alsa/pcm_rate.h>
 
 #include <libswresample/swresample.h>
+#include <libavcodec/avcodec.h>
+#include <libavutil/avutil.h>
 #include <libavutil/channel_layout.h>
 #include <libavutil/opt.h>
 #include <libavutil/mathematics.h>
 #include <libavutil/samplefmt.h>
 
+/* some compatibility wrappers */
+#ifndef AV_VERSION_INT
+#define AV_VERSION_INT(a, b, c) (((a) << 16) | ((b) << 8) | (c))
+#endif
+#ifndef LIBAVCODEC_VERSION_INT
+#define LIBAVCODEC_VERSION_INT  AV_VERSION_INT(LIBAVCODEC_VERSION_MAJOR, \
+LIBAVCODEC_VERSION_MINOR, \
+LIBAVCODEC_VERSION_MICRO)
+#endif
+
+#if LIBAVCODEC_VERSION_INT >= AV_VERSION_INT(61, 3, 100)
+#define USE_CH_LAYOUT
+#endif
 
 static unsigned int filter_size = 16;
 
@@ -95,10 +110,17 @@ static int pcm_src_init(void *obj, snd_pcm_rate_info_t *info)
 		if (!rate->avr)
 			return -ENOMEM;
 
-		av_opt_set_channel_layout(rate->avr, "in_channel_layout",
-					  av_get_default_channel_layout(rate->channels), 0);
-		av_opt_set_channel_layout(rate->avr, "out_channel_layout",
-					  av_get_default_channel_layout(rate->channels), 0);
+		#ifdef USE_CH_LAYOUT
+			AVChannelLayout l;
+			av_channel_layout_default(&l, rate->channels);
+			av_opt_set_chlayout(rate->avr, "in_channel_layout", &l, 0);
+			av_opt_set_chlayout(rate->avr, "out_channel_layout", &l, 0);
+		#else
+			av_opt_set_channel_layout(rate->avr, "in_channel_layout",
+							av_get_default_channel_layout(rate->channels), 0);
+			av_opt_set_channel_layout(rate->avr, "out_channel_layout",
+							av_get_default_channel_layout(rate->channels), 0);
+		#endif
 		av_opt_set_int(rate->avr, "in_sample_rate", rate->in_rate, 0);
 		av_opt_set_int(rate->avr, "out_sample_rate", rate->out_rate, 0);
 		fmt = support_multi_format(rate) ? info->in.format : SND_PCM_FORMAT_S16;
